src/cargo/ops/cargo_compile.rs

```diff
@@ -418,9 +425,37 @@ fn generate_targets<'a>(pkg: &'a Package,
                 find(tests, "test", Target::is_test, test)?;
                 find(benches, "bench", Target::is_bench, &profiles.bench)?;
             }
-            Ok(targets)
+            targets
+        }
+    };
+
+    //Collect the targets that are libraries or have all required features available.
+    let no_features = HashSet::new();
+    let features = features.unwrap_or(&no_features);
+    let mut compatible_targets = Vec::with_capacity(targets.len());
+    for (target, profile) in targets.drain(0..) {
+        if target.is_lib() || match target.required_features() {
+            Some(f) => !f.iter().any(|f| !features.contains(f)),
```

Comment:
Perhaps instead of the double-negation here we could go for `f.iter().all(...)`?